File: browser/src/main/java/org/sqlunet/browser/xn/selector/PosSelectorPointer.java
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>
 */

package org.sqlunet.browser.xn.selector;

import android.os.Parcel;
import android.os.Parcelable;

import org.sqlunet.HasPos;

import androidx.annotation.NonNull;

/**
 * SelectorPointer
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
public class PosSelectorPointer extends SelectorPointer implements HasPos
{
	/**
	 * POS
	 */
	private final Character pos;

	/**
	 * Static field used to regenerate object, individually or as arrays
	 */
	static public final Parcelable.Creator<PosSelectorPointer> CREATOR = new Parcelable.Creator<PosSelectorPointer>()
	{
		@NonNull
		@Override
		public PosSelectorPointer createFromParcel(@NonNull final Parcel parcel)
		{
			return new PosSelectorPointer(parcel);
		}

		@NonNull
		@Override
		public PosSelectorPointer[] newArray(final int size)
		{
			return new PosSelectorPointer[size];
		}
	};

	/**
	 * Constructor from parcel
	 *
	 * @param parcel parcel
	 */
	private PosSelectorPointer(@NonNull Parcel parcel)
	{
		super(parcel);
		String posStr = parcel.readString();
		assert posStr != null;
		this.pos = posStr.charAt(0);
	}

	/**
	 * Constructor
	 *
	 * @param synsetId synset id
	 * @param wordId   word id
	 * @param pos      pos
	 */
	public PosSelectorPointer(final long synsetId, final long wordId, final Character pos)
	{
		super(synsetId, wordId);
		this.pos = pos;
	}

	@NonNull
	@Override
	public Character getPos()
	{
		return this.pos;
	}


	@Override
	public void writeToParcel(@NonNull final Parcel parcel, final int flags)
	{
		super.writeToParcel(parcel, flags);
		parcel.writeString(this.pos.toString());
	}

	@SuppressWarnings("SameReturnValue")
	@Override
	public int describeContents()
	{
		return 0;
	}

	@NonNull
	@Override
	public String toString()
	{
		return super.toString() + ' ' + "pos=" + this.pos;
	}

}

File: browsersn/src/main/java/org/sqlunet/browser/sn/selector/PosSelectorPointer.kt
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>
 */
package org.sqlunet.browser.sn.selector

import android.os.Parcel
import android.os.Parcelable
import org.sqlunet.HasPos

/**
 * SelectorPointer
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
open class PosSelectorPointer : SelectorPointer, HasPos {

    /**
     * POS
     */
    protected val pos: Char
        @JvmName("getPosProperty")
        get

    /**
     * Constructor from parcel
     *
     * @param parcel parcel
     */
    protected constructor(parcel: Parcel) : super(parcel) {
        val posStr = parcel.readString()!!
        pos = posStr[0]
    }

    /**
     * Constructor
     *
     * @param synsetId synset id
     * @param wordId   word id
     * @param pos      pos
     */
    constructor(synsetId: Long, wordId: Long, pos: Char) : super(synsetId, wordId) {
        this.pos = pos
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        super.writeToParcel(parcel, flags)
        parcel.writeString(pos.toString())
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun getPos(): Char {
        return pos
    }

    override fun toString(): String {
        return super.toString() + ' ' + "pos=" + pos
    }

    companion object {

        @JvmField
        val CREATOR = object : Parcelable.Creator<PosSelectorPointer> {

            override fun createFromParcel(parcel: Parcel): PosSelectorPointer {
                return PosSelectorPointer(parcel)
            }

            override fun newArray(size: Int): Array<PosSelectorPointer?> {
                return arrayOfNulls(size)
            }
        }
    }
}
